### PR TITLE
feat: cjs esm setup

### PR DESCRIPTION
**Pull request by Ad96el**
## fixes https://github.com/KILTprotocol/ticket/issues/2758 & https://github.com/KILTprotocol/ticket/issues/2807

I tested the setup in a React (webpack as bundler) and node application. Other bundler are not tested. 
Also, the package is published under the @kiltprotocol namespace.

### CODE REVIEW

**Comment by Ad96el** [comment]
Is there a reason why the tests run not against the mashnet docker image? 
I would like to adjust the CI to use this image instead of the latest-develop. Any concerns?

**Comment by rflechtner** [comment]
> Is there a reason why the tests run not against the mashnet docker image?
> I would like to adjust the CI to use this image instead of the latest-develop. Any concerns?

It is the same image, but hosted on ECR to avoid rate limits. I don't know why this isn't set up as a pull-through cache though, but instead pushed to ECR separately. The latest-develop tag is another issue though, don't see why we wouldn't just test against latest

File: package.json [hunk 1]
```diff
@@ -1,27 +1,21 @@
...
-    "./wellKnownDidConfiguration": {
-      "types": "./wellKnownDidConfiguration/index.d.ts",
-      "default": "./wellKnownDidConfiguration/index.js"
-    },
-    "./types": {
-      "types": "./types/index.d.ts"
-    }
+    "import": "./dist/esm/index.js",
+    "require": "./dist/cjs/index.js",
+    "default": "./dist/cjs/index.js"
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply package.json:10]: Alright. I see the struggle. Would a possible (and maybe simpler) approach would be, to import everything what we want to export in the index.ts?

File: package.json [hunk 2]
```diff
@@ -1,27 +1,21 @@
 {
-  "name": "kilt-extension-api",
+  "name": "@kiltprotocol/kilt-extension-api",
   "version": "0.1.0",
-  "main": "./index.js",
-  "types": "./index.d.ts",
-  "type": "commonjs",
+  "main": "./dist/cjs/index.js",
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply package.json:4]: The default value, if `files` is not specified, is `[*]` which will include all files. Is this not correct? Which files should be excluded then?

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (rflechtner) [reply package.json:10]: We can try, but we'd have to test it carefully. The thing is that different parts of the project are meant to be run in different environments; the well known DID config runs in a node environment (backend), while the getExtension stuff runs in the browser. This is also why exporting everything under the same root is a little weird; there is just no scenario where you would ever need everything. I think we decided against a monorepo, and subpath exports are just the next best thing to isolate code meant for different environments.

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (rflechtner) [reply package.json:4]: some projects do add all source files to the npm package (I guess to help debugging), but it increases package size obviously. We really only need the README, LICENSE, the .js files we build, and the files that are always included (e.g., package.json).

File: package.json [hunk 3]
```diff
@@ -1,27 +1,21 @@
...
+    "require": "./dist/cjs/index.js",
+    "default": "./dist/cjs/index.js"
   },
   "author": "KILT <info@kilt.io>",
   "license": "BSD-4-Clause",
   "scripts": {
     "lint": "eslint --ext .ts .",
-    "build": "tsc -p tsconfig.build.json",
+    "build": "yarn build:cjs && yarn build:esm",
+    "build:cjs": "tsc -p tsconfig.cjs.json",
```
[the remark this answers is not in the document]
Comment (Ad96el) [reply package.json:17]: for my learning: what is the point of having a minimal package.json in the build folders? While debugging, I noticed it also worked without the file

File: package.json [hunk 4]
```diff
@@ -1,29 +1,46 @@
...
+      "import": "./dist/esm/index.js",
+      "require": "./dist/cjs/index.js",
+      "default": "./dist/cjs/index.js"
     },
-    "./wellKnownDidConfiguration": {
-      "types": "./wellKnownDidConfiguration/index.d.ts",
-      "default": "./wellKnownDidConfiguration/index.js"
+    "./wellKnownDidConfiguration" : {
+      "import": "./dist/esm/wellKnownDidConfiguration/index.js",
+      "require": "./dist/cjs/wellKnownDidConfiguration/index.js",
```
Comment (rflechtner) [inline package.json:15]: I think the cjs build should still live in the root, like it did before, for tools that don't understand the exports map yet

Comment (Ad96el) [reply package.json]: cjs build is (thanks 2 u) in the root now :cry:  : [ea72411](https://github.com/KILTprotocol/kilt-extension-api/pull/27/commits/ea7241148c61bc9db532cb40fc3f4054013cfb3a).

File: package.json [hunk 5]
```diff
@@ -1,29 +1,59 @@
...
   "exports": {
     ".": {
-      "types": "./index.d.ts",
+      "import": "./dist/esm/index.js",
+      "require": "./index.js",
       "default": "./index.js"
     },
     "./wellKnownDidConfiguration": {
-      "types": "./wellKnownDidConfiguration/index.d.ts",
+      "import": "./dist/esm/wellKnownDidConfiguration/index.js",
```
Comment (rflechtner) [inline package.json:14]: Can we shorten the path to the esm build to just /esm/ ?

File: tsconfig.cjs.json [hunk 6]
```diff
@@ -0,0 +1,15 @@
+{
+  "extends": "./tsconfig.base.json",
+  "compilerOptions": {
+    "target": "ES2022",                               
+    "module": "CommonJS",                             
+    "declaration": true,                              
+    "outDir": ".",                            
+    "esModuleInterop": true,
```
Comment (rflechtner) [inline tsconfig.cjs.json:8]: We tried to get by without esModuleInterop, is this required?

File: tsconfig.esm.json [hunk 7]
```diff
@@ -0,0 +1,15 @@
+{
+  "extends": "./tsconfig.base.json",
+  "compilerOptions": {
+    "target": "ES2022",                               
+    "module": "ESNext",                             
+    "declaration": true,           
+    "outDir": "./dist/esm",                            
+    "esModuleInterop": true,
```
Comment (rflechtner) [inline tsconfig.esm.json:8]: Same here

File: tsconfig.esm.json [hunk 8]
```diff
@@ -0,0 +1,15 @@
+{
+  "extends": "./tsconfig.base.json",
+  "compilerOptions": {
+    "target": "ES2022",                               
+    "module": "ESNext",                             
+    "declaration": true,           
+    "outDir": "./dist/esm",
```
Comment (rflechtner) [inline tsconfig.esm.json:7]:
```suggestion
    "outDir": "./esm",                            
```

**Review by rflechtner — COMMENTED** [review]
Love it! Maybe we can shorten the esm path though and get rid of esModuleInterop?

File: tsconfig.esm.json (hunk 8 above)
Comment (Ad96el) [reply tsconfig.esm.json]: updated in: [3ae03a2](https://github.com/KILTprotocol/kilt-extension-api/pull/27/commits/3ae03a28dfb736af8ff82c330a70622277a2bdfe)

File: package.json (hunk 5 above)
Comment (Ad96el) [reply package.json]: updated in: [3ae03a2](https://github.com/KILTprotocol/kilt-extension-api/pull/27/commits/3ae03a28dfb736af8ff82c330a70622277a2bdfe)

File: tsconfig.cjs.json (hunk 6 above)
Comment (Ad96el) [reply tsconfig.cjs.json:8]: If I try to build it without I get this error:

```
 error TS1259: Module '""' can only be default-imported using the 'allowSyntheticDefaultImports' flag

10 import yargs from 'yargs/yargs'
          ~~~~~
```

Comment (rflechtner) [reply tsconfig.cjs.json:8]: goddamit! ok than we'll merge it with esModuleInterop set

**Review by rflechtner — APPROVED** [review]
Yeeey, let's merge this!

**Pull request merged**